File: backend/src/models/assistance_request.py
from typing import List, Optional
from pydantic import BaseModel, Field
from bson import ObjectId
from datetime import datetime, timezone


class AssistanceRequest(BaseModel):
    id: Optional[str] = Field(default_factory=lambda: str(ObjectId()), alias="_id")
    title: str = Field(..., description="Title of the assistance request.")
    description: str = Field(..., description="Detailed description of the request.")
    workshop_id: str = Field(..., description="ID of the workshop associated with this request.")
    workers_ids: List[str] = Field(default_factory=list, description="List of worker IDs assigned to this request.")
    is_completed: Optional[bool] = Field(default=False, description="Completion status of the request.")
    creation_date: datetime = Field(default_factory=lambda: datetime.now(timezone.utc), description="Timestamp of when the request was created.")

    class Config:
        allow_population_by_field_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}


class AssistanceRequestCreate(BaseModel):
    title: str = Field(..., description="Title of the assistance request.")
    description: str = Field(..., description="Detailed description of the request.")
    workshop_id: str = Field(..., description="ID of the workshop associated with this request.")
    workers_ids: List[str] = Field(default_factory=list, description="List of worker IDs assigned to this request.")
    is_completed: Optional[bool] = Field(default=False, description="Completion status of the request.")
    creation_date: datetime = Field(default_factory=lambda: datetime.now(timezone.utc), description="Timestamp of when the request was created.")

File: frontend/lib/src/models/assistance_request.dart
class AssistanceRequest {
  final String? id;
   String title;
   String description;
  final String workshopId;
  late final List<String> workersIds;
  bool? isCompleted;
  final DateTime creationDate;

  AssistanceRequest({
    this.id,
    required this.title,
    required this.description,
    required this.workshopId,
    this.workersIds = const [],
    this.isCompleted = false,
    required this.creationDate,
  });

  factory AssistanceRequest.fromJson(Map<String, dynamic> jsonInput) {
    return AssistanceRequest(
      id: jsonInput['_id'] as String?,
      title: jsonInput['title'] as String,
      description: jsonInput['description'] as String,
      workshopId: jsonInput['workshop_id'] as String,
      workersIds: List<String>.from(jsonInput['workers_ids'] ?? []), // Handle null case if 'workers_ids' is not provided
      isCompleted: jsonInput['is_completed'] as bool? ?? false,
      creationDate: DateTime.parse(jsonInput['creation_date'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'title': title,
      'description': description,
      'workshop_id': workshopId,
      'workers_ids': workersIds,
      'is_completed': isCompleted,
      'creation_date': creationDate.toIso8601String(),
    };
  }
}